src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -75,8 +75,51 @@ public ProxyWhitelist(Collection<? extends Whitelist> delegates) {
         reset(delegates);
     }
 
+    /**
+     * This should only be called from {@link ProxyWhitelist#reset(Collection)}. So we can assume that a write 
+     * lock is held.
+     * @param wrapper the wrapper
+     */
+    private void addWrapper(ProxyWhitelist wrapper) {
+        assert wrapper.lock.writeLock().isHeldByCurrentThread();
+        lock.writeLock().lock();
+        try {
+            wrappers.put(wrapper, null);
```

Comment:
This is what cause the problem - the write lock being held was on the wrapper not the delegate.  
This line needs a write lock on `this` (the delegate), the rest of the method only needs a read lock.